libfwupdplugin/fu-edid.c

```diff
@@ -0,0 +1,429 @@
+	return self->pnp_id;
+}
+
+/**
+ * fu_edid_set_pnp_id:
+ * @self: a #FuEdid
+ * @pnp_id: (nullable): string value, or %NULL
+ *
+ * Sets the PNP ID.
+ *
+ * Since: 1.9.6
+ **/
+void
+fu_edid_set_pnp_id(FuEdid *self, const gchar *pnp_id)
+{
+	g_return_if_fail(FU_IS_EDID(self));
+	if (g_strcmp0(self->pnp_id, pnp_id) == 0)
+		return;
+	g_free(self->pnp_id);
+	self->pnp_id = g_strdup(pnp_id);
+}
+
+/**
+ * fu_edid_get_eisa_id:
+ * @self: a #FuEdid
+ *
+ * Gets the PNP ID.
+ *
+ * Returns: string value, or %NULL for unset
+ *
+ * Since: 1.9.6
+ **/
+const gchar *
+fu_edid_get_eisa_id(FuEdid *self)
+{
+	g_return_val_if_fail(FU_IS_EDID(self), NULL);
+	return self->eisa_id;
+}
+
+/**
+ * fu_edid_set_eisa_id:
+ * @self: a #FuEdid
+ * @eisa_id: (nullable): string value, or %NULL
+ *
+ * Sets the PNP ID.
+ *
+ * Since: 1.9.6
+ **/
+void
+fu_edid_set_eisa_id(FuEdid *self, const gchar *eisa_id)
+{
+	g_return_if_fail(FU_IS_EDID(self));
+	if (g_strcmp0(self->eisa_id, eisa_id) == 0)
+		return;
+	g_free(self->eisa_id);
+	self->eisa_id = g_strdup(eisa_id);
+}
+
+/**
+ * fu_edid_get_serial_number:
+ * @self: a #FuEdid
+ *
+ * Gets the serial number.
+ *
+ * Returns: string value, or %NULL for unset
+ *
+ * Since: 1.9.6
+ **/
+const gchar *
+fu_edid_get_serial_number(FuEdid *self)
+{
+	g_return_val_if_fail(FU_IS_EDID(self), NULL);
+	return self->serial_number;
+}
+
+/**
+ * fu_edid_set_serial_number:
+ * @self: a #FuEdid
+ * @serial_number: (nullable): string value, or %NULL
+ *
+ * Sets the serial number.
+ *
+ * Since: 1.9.6
+ **/
+void
+fu_edid_set_serial_number(FuEdid *self, const gchar *serial_number)
+{
+	g_return_if_fail(FU_IS_EDID(self));
+	if (g_strcmp0(self->serial_number, serial_number) == 0)
+		return;
+	g_free(self->serial_number);
+	self->serial_number = g_strdup(serial_number);
+}
+
+/**
+ * fu_edid_get_product_code:
+ * @self: a #FuEdid
+ *
+ * Gets the product code.
+ *
+ * Returns: integer, or 0x0 for unset
+ *
+ * Since: 1.9.6
+ **/
+guint16
+fu_edid_get_product_code(FuEdid *self)
+{
+	g_return_val_if_fail(FU_IS_EDID(self), G_MAXUINT16);
+	return self->product_code;
+}
+
+/**
+ * fu_edid_set_product_code:
+ * @self: a #FuEdid
+ * @product_code: integer, or 0x0 for unset
+ *
+ * Sets the product code.
+ *
+ * Since: 1.9.6
+ **/
+void
+fu_edid_set_product_code(FuEdid *self, guint16 product_code)
+{
+	g_return_if_fail(FU_IS_EDID(self));
+	self->product_code = product_code;
+}
+
+static gboolean
+fu_edid_parse_descriptor(FuEdid *self, GBytes *fw, gsize offset, GError **error)
+{
+	gsize buf2sz = 0;
+	gsize bufsz = 0;
+	const guint8 *buf2;
+	const guint8 *buf = g_bytes_get_data(fw, &bufsz);
+	g_autoptr(GByteArray) st = NULL;
+	g_autofree gchar *val = NULL;
+
+	st = fu_struct_edid_descriptor_parse(buf, bufsz, offset, error);
+	if (st == NULL)
+		return FALSE;
+
+	/* ignore pixel clock data */
+	if (fu_struct_edid_descriptor_get_kind(st) != 0x0 ||
+	    fu_struct_edid_descriptor_get_subkind(st) != 0x0)
+		return TRUE;
```

Comment:
I guess I don't know the purpose of this PR (where it's going), but I would think this is incredibly valuable information.  Whenever you parse an EDID, this is the money stuff.

Reply:
This is for a device where the VID/PID isn't good enough to match the firmware -- so we need something like `VID_1234&PID_5678&PNP_WAC` to match up the firmware.

Reply:
Got it.

I think it would be good to still save it and output it then for firmware-parse, it makes fwupdtool a good Swiss army knife.

Reply:
Right, I agree with you -- but lets wait until there's a usecase so we know what format to export -- e.g. do we want a FuFirmware subclass for each display range, do we just need the yXY primaries, or do we want a MD5 hash of the pixel clock data etc.